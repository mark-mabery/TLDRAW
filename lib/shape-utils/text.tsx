import { uniqueId } from 'utils/utils'
import vec from 'utils/vec'
import { TextShape, ShapeType, FontSize } from 'types'
import { registerShapeUtils } from './index'
import { defaultStyle, getFontStyle, getShapeStyle } from 'lib/shape-styles'
import styled from 'styles'
import state from 'state'
import { useEffect, useRef } from 'react'

// A div used for measurement

if (document.getElementById('__textMeasure')) {
  document.getElementById('__textMeasure').remove()
}

const mdiv = document.createElement('pre')
mdiv.id = '__textMeasure'
mdiv.style.whiteSpace = 'pre'
mdiv.style.width = 'auto'
mdiv.style.border = '1px solid red'
mdiv.style.padding = '4px'
mdiv.style.lineHeight = '1'
mdiv.style.margin = '0px'
mdiv.style.opacity = '0'
mdiv.style.position = 'absolute'
mdiv.style.top = '-500px'
mdiv.style.left = '0px'
mdiv.style.zIndex = '9999'
mdiv.setAttribute('readonly', 'true')
document.body.appendChild(mdiv)

const text = registerShapeUtils<TextShape>({
  isForeignObject: true,
  canChangeAspectRatio: false,
  canEdit: true,

  boundsCache: new WeakMap([]),

  create(props) {
    return {
      id: uniqueId(),
      seed: Math.random(),
      type: ShapeType.Text,
      isGenerated: false,
      name: 'Text',
      parentId: 'page1',
      childIndex: 0,
      point: [0, 0],
      rotation: 0,
      isAspectRatioLocked: false,
      isLocked: false,
      isHidden: false,
      style: defaultStyle,
      text: '',
      scale: 1,
      size: 'auto',
      fontSize: FontSize.Medium,
      ...props,
    }
  },

  render(shape, { isEditing, ref }) {
    const { id, text, style } = shape
    const styles = getShapeStyle(style)
    const font = getFontStyle(shape.fontSize, shape.scale, shape.style)

    const bounds = this.getBounds(shape)

    return (
      <foreignObject
        id={id}
        x={0}
        y={0}
        width={bounds.width}
        height={bounds.height}
        pointerEvents="none"
      >
        {isEditing ? (
          <StyledTextArea
            ref={ref}
            style={{
              font,
              color: styles.stroke,
            }}
            value={text}
            autoComplete="false"
            autoCapitalize="false"
            autoCorrect="false"
            onFocus={(e) => {
              e.currentTarget.select()
              state.send('FOCUSED_EDITING_SHAPE')
            }}
            onBlur={() => {
              state.send('BLURRED_EDITING_SHAPE')
            }}
            onChange={(e) => {
              state.send('EDITED_SHAPE', {
                change: { text: e.currentTarget.value },
              })
            }}
          />
        ) : (
          <StyledText
            style={{
              font,
              color: styles.stroke,
            }}
          >
            {text}
          </StyledText>
        )}
      </foreignObject>
    )
  },

  getBounds(shape) {
    if (!this.boundsCache.has(shape)) {
      mdiv.innerHTML = shape.text || ' ' // + '&nbsp;'
      mdiv.style.font = getFontStyle(shape.fontSize, shape.scale, shape.style)

      const [minX, minY] = shape.point
      const [width, height] = [mdiv.offsetWidth, mdiv.offsetHeight]

      this.boundsCache.set(shape, {
        minX,
        maxX: minX + width,
        minY,
        maxY: minY + height,
        width,
        height,
      })
    }

    return this.boundsCache.get(shape)
  },

  hitTest(shape, test) {
    return true
  },

  transform(shape, bounds, { initialShape, transformOrigin, scaleX, scaleY }) {
    if (shape.rotation === 0 && !shape.isAspectRatioLocked) {
      shape.point = [bounds.minX, bounds.minY]
      shape.scale = initialShape.scale * Math.abs(scaleX)
    } else {
      shape.point = [bounds.minX, bounds.minY]

      shape.rotation =
        (scaleX < 0 && scaleY >= 0) || (scaleY < 0 && scaleX >= 0)
          ? -initialShape.rotation
          : initialShape.rotation

      shape.scale = initialShape.scale * Math.abs(Math.min(scaleX, scaleY))
    }

    return this
  },

  transformSingle(shape, bounds, { initialShape, scaleX }) {
    shape.point = [bounds.minX, bounds.minY]
    shape.scale = initialShape.scale * Math.abs(scaleX)
    return this
  },

  onBoundsReset(shape) {
    shape.size = 'auto'
    return this
  },

  shouldDelete(shape) {
    return shape.text.length === 0
  },
})

export default text

const StyledText = styled('div', {
  width: '100%',
  height: '100%',
  border: 'none',
  padding: '4px',
  whiteSpace: 'pre',
  resize: 'none',
  minHeight: 1,
  minWidth: 1,
  outline: 'none',
  backgroundColor: 'transparent',
  overflow: 'hidden',
  pointerEvents: 'none',
  userSelect: 'none',
})

const StyledTextArea = styled('textarea', {
  width: '100%',
  height: '100%',
  border: 'none',
  padding: '4px',
  whiteSpace: 'pre',
  resize: 'none',
  minHeight: 1,
  minWidth: 1,
  outline: 'none',
  overflow: 'hidden',
  backgroundColor: '$boundsBg',
  pointerEvents: 'all',
})
