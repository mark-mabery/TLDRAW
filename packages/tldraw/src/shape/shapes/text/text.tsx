/* eslint-disable @typescript-eslint/no-non-null-assertion */
import * as React from 'react'
import { HTMLContainer, TLBounds, Utils, ShapeUtil } from '@tldraw/core'
import { Vec } from '@tldraw/vec'
import { getShapeStyle, getFontStyle, defaultStyle } from '~shape/shape-styles'
import { TextShape, TLDrawShapeType, TLDrawToolType, TLDrawMeta } from '~types'
import css from '~styles'
import TextAreaUtils from './text-utils'

const LETTER_SPACING = -1.5

function normalizeText(text: string) {
  return text.replace(/\r?\n|\r/g, '\n')
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
// let melm: any

// function getMeasurementDiv() {
//   // A div used for measurement
//   document.getElementById('__textMeasure')?.remove()

//   const pre = document.createElement('pre')
//   pre.id = '__textMeasure'

//   Object.assign(pre.style, {
//     whiteSpace: 'pre',
//     width: 'auto',
//     border: '1px solid red',
//     padding: '4px',
//     margin: '0px',
//     letterSpacing: `${LETTER_SPACING}px`,
//     opacity: '0',
//     position: 'absolute',
//     top: '-500px',
//     left: '0px',
//     zIndex: '9999',
//     pointerEvents: 'none',
//     userSelect: 'none',
//     alignmentBaseline: 'mathematical',
//     dominantBaseline: 'mathematical',
//   })

//   pre.tabIndex = -1

//   document.body.appendChild(pre)
//   return pre
// }

// if (typeof window !== 'undefined') {
//   melm = getMeasurementDiv()
// }

export const Text = new ShapeUtil<TextShape, HTMLDivElement, TLDrawMeta>(() => ({
  type: TLDrawShapeType.Text,

  toolType: TLDrawToolType.Text,

  isAspectRatioLocked: true,

  isEditableText: true,

  canBind: true,

  defaultProps: {
    id: 'id',
    type: TLDrawShapeType.Text,
    name: 'Text',
    parentId: 'page',
    childIndex: 1,
    point: [0, 0],
    rotation: 0,
    text: ' ',
    style: defaultStyle,
  },

  shouldRender(prev, next): boolean {
    return (
      next.text !== prev.text || next.style.scale !== prev.style.scale || next.style !== prev.style
    )
  },

  Component({ shape, meta, isEditing, isBinding, onShapeChange, onShapeBlur, events }, ref) {
    const rInput = React.useRef<HTMLTextAreaElement>(null)
    const { text, style } = shape
    const styles = getShapeStyle(style, meta.isDarkMode)
    const font = getFontStyle(shape.style)

    const handleChange = React.useCallback(
      (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        onShapeChange?.({ ...shape, text: normalizeText(e.currentTarget.value) })
      },
      [shape]
    )

    const handleKeyDown = React.useCallback(
      (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
        if (e.key === 'Escape') return

        e.stopPropagation()

        if (e.key === 'Tab') {
          e.preventDefault()
          if (e.shiftKey) {
            TextAreaUtils.unindent(e.currentTarget)
          } else {
            TextAreaUtils.indent(e.currentTarget)
          }

          onShapeChange?.({ ...shape, text: normalizeText(e.currentTarget.value) })
        }
      },
      [shape, onShapeChange]
    )

    const handleBlur = React.useCallback(
      (e: React.FocusEvent<HTMLTextAreaElement>) => {
        e.currentTarget.setSelectionRange(0, 0)
        onShapeBlur?.()
      },
      [isEditing, shape]
    )

    const handleFocus = React.useCallback(
      (e: React.FocusEvent<HTMLTextAreaElement>) => {
        if (!isEditing) return
        if (document.activeElement === e.currentTarget) {
          e.currentTarget.select()
        }
      },
      [isEditing]
    )

    const handlePointerDown = React.useCallback(
      (e) => {
        if (isEditing) {
          e.stopPropagation()
        }
      },
      [isEditing]
    )

    React.useEffect(() => {
      if (isEditing) {
        requestAnimationFrame(() => {
          const elm = rInput.current!
          elm.focus()
          elm.select()
        })
      } else {
        const elm = rInput.current!
        elm.setSelectionRange(0, 0)
      }
    }, [isEditing])

    return (
      <HTMLContainer ref={ref} {...events}>
        <div className={wrapper({ isEditing })} onPointerDown={handlePointerDown}>
          <textarea
            className={textArea({ isEditing, isBinding })}
            ref={rInput}
            style={{
              font,
              color: styles.stroke,
            }}
            name="text"
            defaultValue={text}
            tabIndex={-1}
            autoComplete="false"
            autoCapitalize="false"
            autoCorrect="false"
            autoSave="false"
            placeholder=""
            color={styles.stroke}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onChange={handleChange}
            onKeyDown={handleKeyDown}
            onPointerDown={handlePointerDown}
            autoFocus={isEditing}
            readOnly={!isEditing}
            wrap="off"
            dir="auto"
            datatype="wysiwyg"
          />
        </div>
      </HTMLContainer>
    )
  },

  Indicator() {
    return null
  },

  getBounds(shape): TLBounds {
    const ref = this.getRef(shape)
    const elm = ref.current

    const bounds = {
      minX: 0,
      minY: 0,
      maxX: elm?.offsetWidth || 32,
      maxY: elm?.offsetHeight || 32,
      width: elm?.offsetWidth || 32,
      height: elm?.offsetHeight || 32,
    }

    // const bounds = Utils.getFromCache(this.boundsCache, shape, () => {
    //   if (!melm) {
    //     // We're in SSR
    //     return { minX: 0, minY: 0, maxX: 10, maxY: 10, width: 10, height: 10 }
    //   }

    //   melm.innerHTML = `${shape.text}&zwj;`
    //   melm.style.font = getFontStyle(shape.style)

    //   // In tests, offsetWidth and offsetHeight will be 0
    //   const width = melm.offsetWidth || 1
    //   const height = melm.offsetHeight || 1

    //   return {
    //     minX: 0,
    //     maxX: width,
    //     minY: 0,
    //     maxY: height,
    //     width,
    //     height,
    //   }
    // })

    return Utils.translateBounds(bounds, shape.point)
  },

  transform(_shape, bounds, { initialShape, scaleX, scaleY }) {
    const {
      rotation = 0,
      style: { scale = 1 },
    } = initialShape

    const nextScale = scale * Math.abs(Math.min(scaleX, scaleY))

    return {
      point: [bounds.minX, bounds.minY],
      rotation:
        (scaleX < 0 && scaleY >= 0) || (scaleY < 0 && scaleX >= 0) ? -(rotation || 0) : rotation,
      style: {
        ...initialShape.style,
        scale: nextScale,
      },
    }
  },

  transformSingle(_shape, bounds, { initialShape, scaleX, scaleY }) {
    const {
      style: { scale = 1 },
    } = initialShape

    return {
      point: Vec.round([bounds.minX, bounds.minY]),
      style: {
        ...initialShape.style,
        scale: scale * Math.max(Math.abs(scaleY), Math.abs(scaleX)),
      },
    }
  },

  onDoubleClickBoundsHandle(shape) {
    const center = this.getCenter(shape)

    const newCenter = this.getCenter({
      ...shape,
      style: {
        ...shape.style,
        scale: 1,
      },
    })

    return {
      style: {
        ...shape.style,
        scale: 1,
      },
      point: Vec.round(Vec.add(shape.point, Vec.sub(center, newCenter))),
    }
  },

  onStyleChange(shape) {
    const center = this.getCenter(shape)

    this.boundsCache.delete(shape)

    const newCenter = this.getCenter(shape)

    return {
      point: Vec.round(Vec.add(shape.point, Vec.sub(center, newCenter))),
    }
  },
}))

/* -------------------------------------------------- */
/*                       Helpers                      */
/* -------------------------------------------------- */

const wrapper = css({
  width: '100%',
  height: '100%',
  variants: {
    isEditing: {
      false: {
        pointerEvents: 'all',
      },
      true: {
        pointerEvents: 'none',
      },
    },
  },
})

const textArea = css({
  position: 'absolute',
  top: 'var(--tl-padding)',
  left: 'var(--tl-padding)',
  zIndex: 1,
  width: 'calc(100% - (var(--tl-padding) * 2))',
  height: 'calc(100% - (var(--tl-padding) * 2))',
  border: 'none',
  padding: '4px',
  whiteSpace: 'pre',
  alignmentBaseline: 'mathematical',
  dominantBaseline: 'mathematical',
  resize: 'none',
  minHeight: 1,
  minWidth: 1,
  lineHeight: 1.4,
  letterSpacing: LETTER_SPACING,
  outline: 0,
  fontWeight: '500',
  overflow: 'hidden',
  backfaceVisibility: 'hidden',
  display: 'inline-block',
  WebkitUserSelect: 'text',
  WebkitTouchCallout: 'none',
  variants: {
    isBinding: {
      false: {},
      true: {
        background: '$boundsBg',
      },
    },
    isEditing: {
      false: {
        pointerEvents: 'none',
        userSelect: 'none',
        background: 'none',
        WebkitUserSelect: 'none',
      },
      true: {
        pointerEvents: 'all',
        userSelect: 'text',
        background: '$boundsBg',
        WebkitUserSelect: 'text',
      },
    },
  },
})
