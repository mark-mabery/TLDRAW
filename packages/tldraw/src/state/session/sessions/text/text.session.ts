import { TextShape, TLDrawShape, TLDrawStatus } from '~types'
import type { Session } from '~types'
import type { Data } from '~types'
import { TLDR } from '~state/tldr'

export class TextSession implements Session {
  id = 'text'
  status = TLDrawStatus.EditingText
  initialShape: TextShape

  constructor(data: Data, id?: string) {
    this.initialShape = TLDR.getShape(data, id || TLDR.getSelectedIds(data)[0])
  }

  start = (data: Data) => {
    return {
      document: {
        pageStates: {
          [data.appState.currentPageId]: {
            editingId: this.initialShape.id,
          },
        },
      },
    }
  }

  update = (data: Data, text: string) => {
    const {
      initialShape: { id },
    } = this

    let nextShape: TextShape = {
      ...TLDR.getShape<TextShape>(data, id),
      text,
    }

    nextShape = {
      ...nextShape,
      ...TLDR.getShapeUtils(nextShape).onStyleChange(nextShape),
    } as TextShape

    return {
      document: {
        pages: {
          [data.appState.currentPageId]: {
            shapes: {
              [id]: nextShape,
            },
          },
        },
      },
    }
  }

  cancel = (data: Data) => {
    const {
      initialShape: { id },
    } = this

    return {
      document: {
        pages: {
          [data.appState.currentPageId]: {
            shapes: {
              [id]: TLDR.onSessionComplete(data, TLDR.getShape(data, id)),
            },
          },
        },
        pageState: {
          [data.appState.currentPageId]: {
            editingId: undefined,
          },
        },
      },
    }
  }

  complete(data: Data) {
    const { initialShape } = this

    const shape = TLDR.getShape<TextShape>(data, initialShape.id)

    if (shape.text === initialShape.text) return undefined

    return {
      id: 'text',
      before: {
        document: {
          pages: {
            [data.appState.currentPageId]: {
              shapes: {
                [initialShape.id]: initialShape,
              },
            },
          },
          pageState: {
            [data.appState.currentPageId]: {
              editingId: undefined,
            },
          },
        },
      },
      after: {
        document: {
          pages: {
            [data.appState.currentPageId]: {
              shapes: {
                [initialShape.id]: TLDR.onSessionComplete(
                  data,
                  TLDR.getShape(data, initialShape.id)
                ),
              },
            },
          },
          pageState: {
            [data.appState.currentPageId]: {
              editingId: undefined,
            },
          },
        },
      },
    }
  }
}
