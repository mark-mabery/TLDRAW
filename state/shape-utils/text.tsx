import { uniqueId, isMobile } from 'utils'
import vec from 'utils/vec'
import { TextShape, ShapeType, FontSize } from 'types'
import {
  defaultStyle,
  getFontSize,
  getFontStyle,
  getShapeStyle,
} from 'state/shape-styles'
import styled from 'styles'
import state from 'state'
import { registerShapeUtils } from './register'

// A div used for measurement

if (document.getElementById('__textMeasure')) {
  document.getElementById('__textMeasure').remove()
}

// A div used for measurement
const mdiv = document.createElement('pre')
mdiv.id = '__textMeasure'

Object.assign(mdiv.style, {
  whiteSpace: 'pre',
  width: 'auto',
  border: '1px solid red',
  padding: '4px',
  margin: '0px',
  opacity: '0',
  position: 'absolute',
  top: '-500px',
  left: '0px',
  zIndex: '9999',
  pointerEvents: 'none',
})

mdiv.tabIndex = -1

document.body.appendChild(mdiv)

function normalizeText(text: string) {
  return text.replace(/\t/g, '        ').replace(/\r?\n|\r/g, '\n')
}

const text = registerShapeUtils<TextShape>({
  isForeignObject: true,
  canChangeAspectRatio: false,
  canEdit: true,

  boundsCache: new WeakMap([]),

  create(props) {
    return {
      id: uniqueId(),

      type: ShapeType.Text,
      isGenerated: false,
      name: 'Text',
      parentId: 'page1',
      childIndex: 0,
      point: [0, 0],
      rotation: 0,
      isAspectRatioLocked: false,
      isLocked: false,
      isHidden: false,
      style: defaultStyle,
      text: '',
      scale: 1,
      fontSize: FontSize.Medium,
      ...props,
    }
  },

  render(shape, { isEditing, ref }) {
    const { id, text, style } = shape
    const styles = getShapeStyle(style)
    const font = getFontStyle(shape.scale, shape.style)

    const bounds = this.getBounds(shape)

    function handleChange(e: React.ChangeEvent<HTMLTextAreaElement>) {
      state.send('EDITED_SHAPE', {
        change: { text: normalizeText(e.currentTarget.value) },
      })
    }

    function handleKeyDown(e: React.KeyboardEvent) {
      e.stopPropagation()
      if (e.key === 'Tab') {
        e.preventDefault()
      }
    }

    function handleBlur() {
      state.send('BLURRED_EDITING_SHAPE')
    }

    function handleFocus(e: React.FocusEvent<HTMLTextAreaElement>) {
      e.currentTarget.select()
      state.send('FOCUSED_EDITING_SHAPE')
    }

    const fontSize = getFontSize(shape.style.size) * shape.scale
    const gap = fontSize * 0.4

    if (!isEditing) {
      return (
        <g id={id} pointerEvents="none">
          {text.split('\n').map((str, i) => (
            <text
              key={i}
              x={4}
              y={4 + gap / 2 + i * (fontSize + gap)}
              fontFamily="Verveine Regular"
              fontStyle="normal"
              fontWeight="regular"
              fontSize={fontSize}
              width={bounds.width}
              height={bounds.height}
              dominantBaseline="hanging"
            >
              {str}
            </text>
          ))}
        </g>
      )
    }

    return (
      <foreignObject
        id={id}
        x={0}
        y={0}
        width={bounds.width}
        height={bounds.height}
        pointerEvents="none"
      >
        <StyledTextArea
          ref={ref}
          style={{
            font,
            color: styles.stroke,
          }}
          value={text}
          tabIndex={0}
          autoComplete="false"
          autoCapitalize="false"
          autoCorrect="false"
          autoSave="false"
          placeholder=""
          name="text"
          autoFocus={isMobile() ? true : false}
          onFocus={handleFocus}
          onBlur={handleBlur}
          onKeyDown={handleKeyDown}
          onChange={handleChange}
        />
      </foreignObject>
    )
  },

  getBounds(shape) {
    if (!this.boundsCache.has(shape)) {
      mdiv.innerHTML = shape.text + '&zwj;'
      mdiv.style.font = getFontStyle(shape.scale, shape.style)

      const [minX, minY] = shape.point
      const [width, height] = [mdiv.offsetWidth, mdiv.offsetHeight]

      this.boundsCache.set(shape, {
        minX,
        maxX: minX + width,
        minY,
        maxY: minY + height,
        width,
        height,
      })
    }

    return this.boundsCache.get(shape)
  },

  hitTest() {
    return true
  },

  transform(shape, bounds, { initialShape, scaleX, scaleY }) {
    if (shape.rotation === 0 && !shape.isAspectRatioLocked) {
      shape.point = [bounds.minX, bounds.minY]
      shape.scale = initialShape.scale * Math.abs(scaleX)
    } else {
      shape.point = [bounds.minX, bounds.minY]

      shape.rotation =
        (scaleX < 0 && scaleY >= 0) || (scaleY < 0 && scaleX >= 0)
          ? -initialShape.rotation
          : initialShape.rotation

      shape.scale = initialShape.scale * Math.abs(Math.min(scaleX, scaleY))
    }

    return this
  },

  transformSingle(shape, bounds, { initialShape, scaleX }) {
    shape.point = [bounds.minX, bounds.minY]
    shape.scale = initialShape.scale * Math.abs(scaleX)

    return this
  },

  onBoundsReset(shape) {
    const center = this.getCenter(shape)

    this.boundsCache.delete(shape)

    shape.scale = 1

    const newCenter = this.getCenter(shape)

    shape.point = vec.add(shape.point, vec.sub(center, newCenter))

    return this
  },

  applyStyles(shape, style) {
    const center = this.getCenter(shape)

    this.boundsCache.delete(shape)

    Object.assign(shape.style, style)

    const newCenter = this.getCenter(shape)

    shape.point = vec.add(shape.point, vec.sub(center, newCenter))

    return this
  },

  shouldDelete(shape) {
    return shape.text.length === 0
  },
})

export default text

const StyledTextArea = styled('textarea', {
  zIndex: 1,
  width: '100%',
  height: '100%',
  border: 'none',
  padding: '4px',
  whiteSpace: 'pre',
  resize: 'none',
  minHeight: 1,
  minWidth: 1,
  outline: 0,
  backgroundColor: '$boundsBg',
  overflow: 'hidden',
  pointerEvents: 'all',
  backfaceVisibility: 'hidden',
  display: 'inline-block',
  userSelect: 'text',
  WebkitUserSelect: 'text',
  WebkitTouchCallout: 'none',
})
