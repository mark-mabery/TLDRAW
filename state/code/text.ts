import CodeShape from './index'
import { uniqueId } from 'utils/utils'
import { TextShape, ShapeProps, ShapeType } from 'types'
import { defaultStyle } from 'state/shape-styles'
import { getShapeUtils } from 'state/shape-utils'

/* ----------------- Start Copy Here ---------------- */

export default class Text extends CodeShape<TextShape> {
  constructor(props = {} as ShapeProps<TextShape>) {
    super({
      id: uniqueId(),

      parentId: (window as any).currentPageId,
      type: ShapeType.Text,
      isGenerated: true,
      name: 'Text',
      childIndex: 0,
      point: [0, 0],
      rotation: 0,
      isAspectRatioLocked: false,
      isLocked: false,
      isHidden: false,
      text: 'Text',
      scale: 1,
      ...props,
      style: {
        ...defaultStyle,
        ...props.style,
      },
    })
  }

  /**
   * The text shape's text content.
   *
   * ```ts
   * const shapeText = shape.text
   *
   * shape.text = "Hello world!"
   * ```
   */
  get text(): string {
    return this.shape.text
  }

  set text(text: string) {
    getShapeUtils(this.shape).setProperty(this.shape, 'text', text)
  }

  /**
   * The text's scale.
   *
   * ```ts
   * const shapeScale = shape.scale
   *
   * shape.scale = 2
   * ```
   */
  get scale(): number {
    return this.shape.scale
  }

  set scale(scale: number) {
    getShapeUtils(this.shape).setProperty(this.shape, 'scale', scale)
  }
}
