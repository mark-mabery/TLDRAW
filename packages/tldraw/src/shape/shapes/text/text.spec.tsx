import { Text } from './text'

describe('Text shape', () => {
  it('Creates a shape', () => {
    expect(Text.create({ id: 'text' })).toMatchSnapshot('text')
  })
})
