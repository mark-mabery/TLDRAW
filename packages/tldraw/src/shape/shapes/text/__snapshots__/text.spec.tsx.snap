// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text shape Creates a shape: text 1`] = `
Object {
  "childIndex": 1,
  "id": "text",
  "name": "Text",
  "parentId": "page",
  "point": Array [
    0,
    0,
  ],
  "rotation": 0,
  "style": Object {
    "color": "Black",
    "dash": "Draw",
    "isFilled": false,
    "size": "Small",
  },
  "text": " ",
  "type": "text",
}
`;
